File: example/src/test/java/ru/virgil/example/FileImageApiTest.java
package ru.virgil.example;

import com.google.common.truth.Truth;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import ru.virgil.example.image.ImageService;
import ru.virgil.example.image.PrivateImageFile;
import ru.virgil.example.image.PrivateImageFileDto;
import ru.virgil.example.user.UserDetails;
import ru.virgil.example.util.security.user.WithMockFirebaseUser;
import ru.virgil.utils.fluent_request.RequestUtil;
import ru.virgil.utils.image.FileTypeService;
import ru.virgil.utils.image.ImageMockService;

import java.util.UUID;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@RequiredArgsConstructor(onConstructor_ = @Autowired)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class FileImageApiTest {

    private final ImageMockService<UserDetails, PrivateImageFile> imageMockService;
    private final ImageService imageService;
    private final RequestUtil requestUtil;
    private final FileTypeService fileTypeService;

    @Test
    @WithMockFirebaseUser
    void postPrivateImage() throws Exception {
        PrivateImageFileDto privateImageFileDto = (PrivateImageFileDto) requestUtil.postMultipart("/image/private")
                .file(imageMockService.mockAsMultipart())
                .receive(PrivateImageFileDto.class)
                .and()
                .expect(status().isOk());
        Truth.assertThat(privateImageFileDto).isNotNull();
    }

    @Test
    @WithMockFirebaseUser
    public void getPrivateImage() throws Exception {
        PrivateImageFileDto privateImageFileDto = (PrivateImageFileDto) requestUtil.postMultipart("/image/private")
                .file(imageMockService.mockAsMultipart())
                .receive(PrivateImageFileDto.class)
                .and()
                .expect(status().isOk());
        Truth.assertThat(privateImageFileDto).isNotNull();
        byte[] byteArray = (byte[]) requestUtil.get("/image/private/%s".formatted(privateImageFileDto.getUuid()))
                .receiveAsBytes()
                .and()
                .expect(status().isOk());
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/");
    }

    @Test
    @WithMockFirebaseUser
    public void getProtectedImage() throws Exception {
        byte[] byteArray = (byte[]) requestUtil.get("/image/protected/image.jpg")
                .receiveAsBytes()
                .and()
                .expect(status().isOk());
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/");
    }

    @Test
    public void getPublicImage() throws Exception {
        byte[] byteArray = (byte[]) requestUtil.get("/image/public/image.jpg")
                .receiveAsBytes()
                .and()
                .expect(status().isOk());
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/");
    }

    @Test
    @WithMockFirebaseUser
    void getNotExisting() throws Exception {
        requestUtil.get("/image/public/not_existing.jpg")
                .and()
                .expect(status().isNotFound());
        requestUtil.get("/image/protected/not_existing.jpg")
                .and()
                .expect(status().isNotFound());
        requestUtil.get("/image/protected/%s".formatted(UUID.randomUUID()))
                .and()
                .expect(status().isNotFound());
    }

    @AfterAll
    public void cleanUp() {
        // TODO при запуске _всех_ тестов не срабатывает, при запуска _одного_ – срабатывает
        imageService.cleanFolders();
    }
}

File: example/src/test/java/ru/virgil/example/FileImageApiTest.kt
package ru.virgil.example

import com.google.common.truth.Truth
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.ComponentScan
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import ru.virgil.example.image.ImageService
import ru.virgil.example.image.PrivateImageFile
import ru.virgil.example.image.PrivateImageFileDto
import ru.virgil.example.user.UserDetails
import ru.virgil.example.util.security.user.WithMockFirebaseUser
import ru.virgil.test_utils.fluent_request.RequestUtil
import ru.virgil.utils.image.FileTypeService
import ru.virgil.utils.image.ImageMockService
import java.util.*

@SpringBootTest
@AutoConfigureMockMvc
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@ComponentScan("ru.virgil.test_utils")
class FileImageApiTest @Autowired constructor(
    private val imageMockService: ImageMockService<UserDetails, PrivateImageFile>,
    private val imageService: ImageService,
    private val requestUtil: RequestUtil,
    private val fileTypeService: FileTypeService,
) {

    @Test
    @WithMockFirebaseUser
    @Throws(Exception::class)
    fun postPrivateImage() {
        val privateImageFileDto = requestUtil.postMultipart("/image/private")
            .file(imageMockService.mockAsMultipart())
            .receive(PrivateImageFileDto::class.java)
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as PrivateImageFileDto
        Truth.assertThat(privateImageFileDto).isNotNull()
    }

    @Throws(Exception::class)
    @WithMockFirebaseUser
    @Test
    fun getPrivateImage(): Unit {
        val privateImageFileDto = requestUtil.postMultipart("/image/private")
            .file(imageMockService.mockAsMultipart())
            .receive(PrivateImageFileDto::class.java)
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as PrivateImageFileDto
        Truth.assertThat(privateImageFileDto).isNotNull()
        val byteArray = requestUtil["/image/private/%s".formatted(privateImageFileDto.uuid)]
            .receiveAsBytes()
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as ByteArray
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/")
    }

    @Throws(Exception::class)
    @WithMockFirebaseUser
    @Test
    fun getProtectedImage(): Unit {
        val byteArray = requestUtil.get("/image/protected/image.jpg")
            .receiveAsBytes()
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as ByteArray
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/")
    }

    @Throws(Exception::class)
    @Test
    fun getPublicImage(): Unit {
        val byteArray = requestUtil.get("/image/public/image.jpg")
            .receiveAsBytes()
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as ByteArray
        Truth.assertThat(fileTypeService.getImageMimeType(byteArray)).contains("image/")
    }

    @Throws(Exception::class)
    @WithMockFirebaseUser
    @Test
    fun getNotExisting(): Unit {
        requestUtil.get("/image/public/not_existing.jpg")
            .and()
            .expect(MockMvcResultMatchers.status().isNotFound)
        requestUtil.get("/image/protected/not_existing.jpg")
            .and()
            .expect(MockMvcResultMatchers.status().isNotFound)
        requestUtil.get("/image/protected/%s".format(UUID.randomUUID()))
            .and()
            .expect(MockMvcResultMatchers.status().isNotFound)
    }

    /**
     * Сделал удаление после разрушения компонента [ImageMockService], может удалить нужные пользвоателю картинки,
     * костыль.
     */
    @AfterAll
    fun cleanUp() {
        // TODO при запуске _всех_ тестов не срабатывает, при запуска _одного_ – срабатывает
    }
}
